### PR TITLE
Update "Argon" skin gameplay counters to new design

### LINKED ISSUE #18888: Update gameplay HUD visuals
New HUD designs have been available for some time now. This involves applying the new layout, and also updating pieces which feel pretty jank currently (like the song progress display). May require some extra design effort.

### PR DESCRIPTION
**Pull request by frenzibyte**
- Closes #18888 
- [x] Depends on ppy/osu-resources#292
- [x] Depends on ppy/osu-resources#295
- [x] Depends on resource bump
- [x] Depends on #25402 (for size serialisation)

This is roughly a big one, containing all the changes necessary to bring the HUD to a similar state with the design (at least for the top side of the screen).

I didn't attempt to split this as this mostly contains new components with as less complexity as possible, but I've organised the changes to allow commit-by-commit reviewing for convenience).

https://github.com/ppy/osu/assets/22781491/c043889d-050a-44e3-9a4f-99ec21755c5e

Notes:
 - The combo/accuracy counters should probably be set to fixed-width, but I've kept them as-is to match design, leaving it to others for tweaking it.

 - The argon health display's width is slightly less than the component's actual dimensions, due to the little white bar preceding the health bar.

 - The font size specifications are intentionally different from the reference designs. To be specific, they're set to 1.2x the reference specifications, to remain visually matching with the designs (see https://github.com/ppy/osu-framework/issues/3271 & https://github.com/ppy/osu-framework/pull/5123).

 - The counters could be given a nice blue-ish tint by bringing the wedges to be on top of them:
   
   https://github.com/ppy/osu/assets/22781491/2c854ce4-d78f-492c-a964-6a3ed294f713
 
 - The glyph scale factors has been tweaked to match the reference design at 1x drawable scale (settling at 0.125x).

 - I've intentionally exposed `ArgonWedgePiece` to allow for users to utilise the wedges in their custom "Argon" skins, seeing that there's no harm in doing so.

### CODE REVIEW

File: osu.Game/Screens/Play/HUD/ArgonScoreCounter.cs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+                    Shadow = false;
+                    UseFullGlyphHeight = false;
+                }
+
+                [BackgroundDependencyLoader]
+                private void load(ISkinSource skin)
+                {
+                    Font = new FontUsage(@"argon-score", 1, fixedWidth: true);
+                    Spacing = new Vector2(-2, 0);
```
Comment (peppy) [inline osu.Game/Screens/Play/HUD/ArgonScoreCounter.cs:90]: This could probably be avoided if the sprites didn't have huge whitespace padding, but I guess it's fine for now. We can readjust if/when this gets skin support.

**Comment by peppy** [comment]
### Score needs to expand to the left, not right.

Case 1:

CleanShot 2023-10-30 at 07 44 54

Case 2 (looks silly):

https://github.com/ppy/osu/assets/191335/06c0f904-4df0-4501-8526-cb441e1af061

### Curves don't match

CleanShot 2023-10-30 at 07 47 23

### All the wedge components should be local to the skin and not exposed via the editor

Having an `ArgonWedgePiece` in the skin editor is fine. Having 5 is not. They should be locally defined and/or hidden.

File: osu.Game/Screens/Play/HUD/ArgonHealthDisplay.cs [hunk 2]
```diff
@@ -28,15 +29,15 @@ public partial class ArgonHealthDisplay : HealthDisplay, ISerialisableDrawable
...
         {
             MinValue = 0,
             MaxValue = 64,
             Precision = 1
         };
 
         [SettingSource("Bar length")]
-        public BindableFloat BarLength { get; } = new BindableFloat(0.98f)
+        public BindableFloat BarLength { get; } = new BindableFloat(0.35f)
         {
```
Comment (peppy) [inline osu.Game/Screens/Play/HUD/ArgonHealthDisplay.cs:41]: I don't think these should be changed as they were sane defaults for a more broad use case.

File: osu.Game/Screens/Play/HUD/ArgonHealthDisplay.cs [hunk 3]
```diff
@@ -100,36 +101,45 @@ private void load()
...
+                        {
+                            BarColour = Color4.White,
+                            GlowColour = main_bar_glow_colour,
+                            Blending = BlendingParameters.Additive,
+                            Colour = ColourInfo.GradientHorizontal(Color4.White.Opacity(0.8f), Color4.White),
+                            PathRadius = 40f,
+                            // Kinda hacky, but results in correct positioning with increased path radius.
+                            Margin = new MarginPadding(-30f),
+                            GlowPortion = 0.9f,
+                        },
```
Comment (peppy) [inline osu.Game/Screens/Play/HUD/ArgonHealthDisplay.cs:131]: Please implement the line outside of this class and revert these changes.

File: osu.Game/Screens/Play/HUD/GameplayPerformancePointsCounter.cs [hunk 4]
```diff
@@ -0,0 +1,186 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+#nullable disable
```
Comment (peppy) [inline osu.Game/Screens/Play/HUD/GameplayPerformancePointsCounter.cs:4]: Can you split this into its own PR?

**Review by peppy — DISMISSED** [review]

**Comment by frenzibyte** [comment]
Blocking this PR pending feedback on a new design from @arflyte.

File: osu.Game/Screens/Play/HUD/GameplayPerformancePointsCounter.cs (hunk 4 above)
Comment (peppy) [reply osu.Game/Screens/Play/HUD/GameplayPerformancePointsCounter.cs:4]: @frenzibyte in the mean time, you could still look at this one.

File: osu.Game/Screens/Play/HUD/ArgonHealthDisplay.cs (hunk 3 above)
Comment (frenzibyte) [reply osu.Game/Screens/Play/HUD/ArgonHealthDisplay.cs:131]: I was going to bring this back in a separate class, but after applying https://github.com/ppy/osu/pull/25226#discussion_r1375789990 as well, I think it's better to just remove it altogether and have the health bar fill up the entire screen...

CleanShot 2023-11-05 at 01 34 10

**Comment by frenzibyte** [comment]
> ### All the wedge components should be local to the skin and not exposed via the editor
> Having an `ArgonWedgePiece` in the skin editor is fine. Having 5 is not. They should be locally defined and/or hidden.

If they should not be exposed to the editor, then they would have to be either ingrained within the respective counters as a separate grouping component itself (e.g. `ArgonScoreAccuracyCounters`), or placed off statically without the user being able to remove them (which is feasible with the `ISerialisableDrawable.IsEditable` property).

**Comment by frenzibyte** [comment]
For anyone concerned, the visuals of this PR has changed as part of applying the above requested changes.

Before:

CleanShot 2023-11-05 at 02 10 49

After:

CleanShot 2023-11-05 at 01 58 04

(PP counter has been moved to another PR to reduce complexity)

File: osu.Game/Screens/Play/HUD/ArgonHealthDisplay.cs (hunk 2 above)
Comment (frenzibyte) [reply osu.Game/Screens/Play/HUD/ArgonHealthDisplay.cs]: Resolved in https://github.com/ppy/osu/pull/25226/commits/91cf237fc12fcbecc6d1345310717407f559a32e

File: osu.Game/Screens/Play/HUD/ArgonHealthDisplay.cs (hunk 3 above)
Comment (frenzibyte) [reply osu.Game/Screens/Play/HUD/ArgonHealthDisplay.cs]: Line removed in https://github.com/ppy/osu/pull/25226/commits/1c844a155e97059e23cdae6470401d217fa88db3

File: osu.Game/Screens/Play/HUD/GameplayPerformancePointsCounter.cs (hunk 4 above)
Comment (frenzibyte) [reply osu.Game/Screens/Play/HUD/GameplayPerformancePointsCounter.cs]: PP counter reverted in https://github.com/ppy/osu/pull/25226/commits/58a830fc5b2bd749ebccd7cf1b5e0f12b6f4c117, https://github.com/ppy/osu/pull/25226/commits/a23dfbeab55bb3e195407bcf436444f150d5a355, https://github.com/ppy/osu/pull/25226/commits/6c3169a0ed6e9f0f9e1e7d668993e04861f63540, and https://github.com/ppy/osu/pull/25226/commits/77f5a4cdf53fca660a71140b84c9897439ddbf2e. Now reside in https://github.com/ppy/osu/pull/25362 instead.

**Comment by peppy** [comment]
Just to convey information from discord, @arflyte has updated the design somewhat (https://www.figma.com/file/Tu1C6CR2XOdhm7kxq1JQNj/Gameplay-Update?type=design&node-id=0-1&mode=design) and it should be in a good enough state to refactor this PR to match the new design.

**Comment by frenzibyte** [comment]
Refactor is still WIP (I'll request review when it is done). Currently pending a place to put the accuracy counter in the new layout...
CleanShot 2023-11-07 at 02 05 00

Have brought it up in [discord](https://discord.com/channels/188630481301012481/188630652340404224/1171207909011636275).

**Comment by frenzibyte** [comment]
I've added the accuracy to the top right of the screen in accordance with [current design directions](https://discord.com/channels/188630481301012481/188630652340404224/1171471727881109609) just to keep the ball rolling. Note that this requires pushing some assets to `osu-resources` first, but I'm not doing it yet until flyte has finalised the glyphs.

Preview:

CleanShot 2023-11-08 at 02 04 06

**Comment by peppy** [comment]
> but I'm not doing it yet until flyte has finalised the glyphs.

Can't you push these so this branch can be tested / move forward, but leave in a blocked state until finalised?

**Comment by frenzibyte** [comment]
> > but I'm not doing it yet until flyte has finalised the glyphs.
> 
> Can't you push these so this branch can be tested / move forward, but leave in a blocked state until finalised?

Wasn't sure about pushing something that's not looked at by flyte, but sure, can do.

File: osu.Game/Screens/Play/HUD/ArgonCounterTextComponent.cs [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+        public Container NumberContainer { get; private set; }
+
+        public LocalisableString Text
+        {
+            get => textPart.Text;
+            set
+            {
+                int remainingCount = RequiredDisplayDigits.Value - value.ToString().Count(char.IsDigit);
+                string remainingText = remainingCount > 0 ? new string('#', remainingCount) : string.Empty;
```
Comment (peppy) [inline osu.Game/Screens/Play/HUD/ArgonCounterTextComponent.cs:38]: We might need to reconsider these implementation from a performance perspective in the next profiling round. These string functions can be quite expensive when this text is being updated every update frame.

Comment (frenzibyte) [reply osu.Game/Screens/Play/HUD/ArgonCounterTextComponent.cs:38]: I've spent some time benchmarking this with various methods of allocating a string, and ended up with the following results:
| Method                         | Count | Mean         | Error       | StdDev      | Gen0     | Gen1    | Allocated |
|------------------------------- |------ |-------------:|------------:|------------:|---------:|--------:|----------:|
| AllocateWithStringConstructor  | 1     |     198.7 ns |     2.44 ns |     2.28 ns |   0.2487 |       - |     520 B |
| AllocateWithStringBuilder      | 1     |     200.3 ns |     3.64 ns |     3.04 ns |   0.2487 |       - |     520 B |
| AllocateWithStringCreateMethod | 1     |     290.3 ns |     2.06 ns |     1.72 ns |   0.1988 |       - |     416 B |
| AllocateAndCache | 1     |    378.5 ns |   7.45 ns |   8.28 ns |  0.6733 |   1.38 KB |
| AllocateWithStringConstructor  | 10    |   1,667.8 ns |    33.02 ns |    46.30 ns |   1.9341 |       - |    4048 B |
| AllocateWithStringBuilder      | 10    |   1,629.6 ns |    16.52 ns |    14.64 ns |   1.9341 |       - |    4048 B |
| AllocateWithStringCreateMethod | 10    |   2,790.3 ns |    25.84 ns |    24.17 ns |   1.8845 |       - |    3944 B |
| AllocateAndCache | 10    |    920.1 ns |  17.87 ns |  16.71 ns |  1.0166 |   2.08 KB |
| AllocateWithStringConstructor  | 100   |  16,720.6 ns |   320.20 ns |   393.23 ns |  18.7988 |       - |   39328 B |
| AllocateWithStringBuilder      | 100   |  16,475.0 ns |   249.08 ns |   286.85 ns |  18.7988 |       - |   39328 B |
| AllocateWithStringCreateMethod | 100   |  29,170.3 ns |   362.28 ns |   302.52 ns |  18.7378 |       - |   39224 B |
| AllocateAndCache | 100   |  6,215.7 ns | 124.16 ns | 174.06 ns |  4.4479 |   9.11 KB |
| AllocateWithStringConstructor  | 1000  | 175,346.1 ns | 2,218.49 ns | 2,075.18 ns | 133.5449 | 35.1563 |  392128 B |
| AllocateWithStringBuilder      | 1000  | 172,832.6 ns | 1,270.34 ns |   991.80 ns | 133.0566 | 34.9121 |  392128 B |
| AllocateWithStringCreateMethod | 1000  | 291,833.8 ns | 5,300.95 ns | 4,958.51 ns |  82.0313 | 33.6914 |  392024 B |
| AllocateAndCache | 1000  | 60,663.7 ns | 875.24 ns | 730.86 ns | 38.4521 |  79.42 KB |

<details>
<summary>code</summary>

```csharp
// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
// See the LICENCE file in the repository root for full licence text.

using System.Collections.Generic;
using System.Text;
using BenchmarkDotNet.Attributes;

namespace osu.Game.Benchmarks
{
    public class BenchmarkStrings : BenchmarkTest
    {
        [Params(1, 10, 100, 1000)]
        public int Count { get; set; }

        [Benchmark]
        public string[] AllocateWithStringConstructor()
        {
            StringBuilder builder = new StringBuilder();
            string[] array = new string[Count * 10];

            for (int i = 0; i < Count; i++)
            {
                for (int j = 0; j < 10; j++)
                {
                    builder.Clear();
                    builder.Append('#', j);
                    array[i * 10 + j] = builder.ToString();
                }
            }

            return array;
        }

        [Benchmark]
        public string[] AllocateWithStringBuilder()
        {
            StringBuilder builder = new StringBuilder();
            string[] array = new string[Count * 10];

            for (int i = 0; i < Count; i++)
            {
                for (int j = 0; j < 10; j++)
                {
                    builder.Clear();
                    builder.Append('#', j);
                    array[i * 10 + j] = builder.ToString();
                }
            }

            return array;
        }

        [Benchmark]
        public string[] AllocateWithStringCreateMethod()
        {
            string[] array = new string[Count * 10];

            for (int i = 0; i < Count; i++)
            {
                for (int j = 0; j < 10; j++)
                    array[i * 10 + j] = string.Create(j, '#', (span, state) => span.Fill(state));
            }

            return array;
        }

        [Benchmark]
        public string[] AllocateAndCache()
        {
            Dictionary<int, string> dictionary = new Dictionary<int, string>();
            string[] array = new string[Count * 10];

            for (int i = 0; i < Count; i++)
            {
                for (int j = 0; j < 10; j++)
                {
                    if (!dictionary.TryGetValue(j, out string? value))
                        dictionary[j] = value = new string('#', j);

                    array[i * 10 + j] = value;
                }
            }

            return array;
        }
    }
}
```
</details>

With allocating strings of length 0-10 constant characters over thousand iterations, time sits around 0.175ms, so as far as I understand, it takes around ~0.000175ms on average to allocate a single string per frame, which should be safe enough to not have to add some sort of caching for, but you're more knowledgable in this area than I do.

Comment (peppy) [reply osu.Game/Screens/Play/HUD/ArgonCounterTextComponent.cs:38]: it's not cpu but memory concerns (which leads to GC etc.) so look at the "allocated" column if anything.

**Review by bdach — DISMISSED** [review]

**Review by bdach — APPROVED** [review]
